src/compiler/checker.ts

```diff
@@ -11882,8 +11882,18 @@ namespace ts {
                 if (symbol.flags & SymbolFlags.Property &&
                     (expr.kind === SyntaxKind.PropertyAccessExpression || expr.kind === SyntaxKind.ElementAccessExpression) &&
                     (expr as PropertyAccessExpression | ElementAccessExpression).expression.kind === SyntaxKind.ThisKeyword) {
-                    const func = getContainingFunction(expr);
-                    return !(func && func.kind === SyntaxKind.Constructor && func.parent === symbol.valueDeclaration.parent);
+                    return !isInConstructor(getContainingFunction(expr));
+                    function isInConstructor(func: FunctionLikeDeclaration) {

```

Comment:
I think this would read better closer to the old style -- just `const func = getContainingFunction(expr); if (!func) ...`.

The trailing nested function strikes me as a Haskell-ism.